File: ios/Gogai/Stores/GroupStore.swift
import Foundation

final class GroupStore: ObservableObject {
    @Published var groups: [Group] = []
    @Published private(set) var isLoading = false
    @Published var error: Error?
    /// シークレットグループの表示フラグ（保存されない、バックグラウンド復帰でリセット）
    @Published var showSecretGroups = false
    /// 折りたたまれているグループIDのセット（未含有 = 展開済み）
    @Published private var collapsedGroupIds: Set<Int> = []

    private let cache: AppCache
    private var client: (any APIClientProtocol)?

    init(cache: AppCache = .shared) {
        self.cache = cache
        // 起動時にキャッシュからグループ一覧を読み込む
        self.groups = cache.loadGroups()
    }

    func configure(with client: any APIClientProtocol) {
        self.client = client
    }

    var visibleGroups: [Group] {
        showSecretGroups ? groups : groups.filter { !$0.isSecret }
    }

    /// 非表示にすべきシークレットグループ所属フィードの ID 集合。
    /// シークレット表示中（showSecretGroups=true）は隠すものがないため空集合。
    func secretFeedIds(in feeds: [Feed]) -> Set<Int> {
        guard !showSecretGroups else { return [] }
        let secretGroupIds = Set(groups.filter { $0.isSecret }.map { $0.id })
        guard !secretGroupIds.isEmpty else { return [] }
        return Set(feeds.compactMap { feed in
            guard let gid = feed.group_id, secretGroupIds.contains(gid) else { return nil }
            return feed.id
        })
    }

    @MainActor
    func isExpanded(id: Int) -> Bool {
        !collapsedGroupIds.contains(id)
    }

    @MainActor
    func toggleExpanded(id: Int) {
        if collapsedGroupIds.contains(id) {
            collapsedGroupIds.remove(id)
        } else {
            collapsedGroupIds.insert(id)
        }
    }

    @MainActor
    func fetchGroups() async {
        guard let client else { return }
        isLoading = true
        defer { isLoading = false }
        do {
            groups = try await GroupRepository(client: client).fetchAll()
            cache.saveGroups(groups)
        } catch {
            self.error = error
        }
    }

    @MainActor
    func createGroup(name: String) async throws {
        guard let client else { return }
        let newGroup = try await GroupRepository(client: client).create(name: name)
        groups.append(newGroup)
    }

    @MainActor
    func updateGroup(id: Int, name: String, isSecret: Int? = nil) async throws {
        guard let client else { return }
        let updated = try await GroupRepository(client: client).update(id: id, name: name, isSecret: isSecret)
        if let idx = groups.firstIndex(where: { $0.id == id }) {
            groups[idx] = updated
        }
    }

    @MainActor
    func deleteGroup(id: Int) async throws {
        guard let client else { return }
        try await GroupRepository(client: client).delete(id: id)
        groups.removeAll { $0.id == id }
        collapsedGroupIds.remove(id)
    }

    @MainActor
    func refreshGroup(id: Int) async throws -> RefreshResult {
        guard let client else { throw APIError.invalidURL }
        return try await GroupRepository(client: client).refresh(id: id)
    }

    @MainActor
    func reorderGroups(from source: IndexSet, to destination: Int) async throws {
        guard let client else { return }
        groups.move(fromOffsets: source, toOffset: destination)
        let ids = groups.map { $0.id }
        try await GroupRepository(client: client).reorder(ids: ids)
    }
}

File: android/app/src/main/kotlin/com/mtkg/gogai/store/GroupStore.kt
package com.mtkg.gogai.store

import com.mtkg.gogai.cache.AppCache
import com.mtkg.gogai.model.Feed
import com.mtkg.gogai.model.Group
import com.mtkg.gogai.model.RefreshResult
import com.mtkg.gogai.network.ApiClient
import com.mtkg.gogai.network.ApiException
import com.mtkg.gogai.repository.GroupRepository
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

/// グループ一覧・シークレット表示・展開状態を保持する Store（iOS GroupStore の移植）
class GroupStore(private val cache: AppCache) {
    private val _groups = MutableStateFlow<List<Group>>(cache.loadGroups())
    val groups: StateFlow<List<Group>> = _groups.asStateFlow()

    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()

    private val _error = MutableStateFlow<Throwable?>(null)
    val error: StateFlow<Throwable?> = _error.asStateFlow()

    /// シークレットグループの表示フラグ（保存されない、バックグラウンド復帰でリセット）
    private val _showSecretGroups = MutableStateFlow(false)
    val showSecretGroups: StateFlow<Boolean> = _showSecretGroups.asStateFlow()

    /// 折りたたまれているグループIDのセット（未含有 = 展開済み）
    private val _collapsedGroupIds = MutableStateFlow<Set<Int>>(emptySet())

    private var client: ApiClient? = null

    fun configure(client: ApiClient) {
        this.client = client
    }

    val visibleGroups: List<Group>
        get() = if (_showSecretGroups.value) _groups.value else _groups.value.filterNot { it.isSecret }

    fun setShowSecretGroups(value: Boolean) {
        _showSecretGroups.value = value
    }

    /// バックグラウンド復帰時にシークレット表示状態をリセットする。
    /// 呼び出しは Phase 3 の UI 側（ProcessLifecycle 相当）で行う。
    fun resetSecretVisibility() {
        _showSecretGroups.value = false
    }

    /// 非表示にすべきシークレットグループ所属フィードの ID 集合。
    /// シークレット表示中（showSecretGroups=true）は隠すものがないため空集合。
    fun secretFeedIds(feeds: List<Feed>): Set<Int> {
        if (_showSecretGroups.value) return emptySet()
        val secretGroupIds = _groups.value.filter { it.isSecret }.map { it.id }.toSet()
        if (secretGroupIds.isEmpty()) return emptySet()
        return feeds.mapNotNull { feed ->
            val gid = feed.group_id
            if (gid != null && secretGroupIds.contains(gid)) feed.id else null
        }.toSet()
    }

    fun isExpanded(id: Int): Boolean = !_collapsedGroupIds.value.contains(id)

    fun toggleExpanded(id: Int) {
        _collapsedGroupIds.value = if (_collapsedGroupIds.value.contains(id)) {
            _collapsedGroupIds.value - id
        } else {
            _collapsedGroupIds.value + id
        }
    }

    suspend fun fetchGroups() {
        val client = this.client ?: return
        _isLoading.value = true
        try {
            val fetched = GroupRepository(client).fetchAll()
            _groups.value = fetched
            cache.saveGroups(fetched)
        } catch (e: CancellationException) {
            throw e
        } catch (e: Exception) {
            _error.value = e
        } finally {
            _isLoading.value = false
        }
    }

    suspend fun createGroup(name: String) {
        val client = this.client ?: return
        val newGroup = GroupRepository(client).create(name)
        _groups.value = _groups.value + newGroup
    }

    suspend fun updateGroup(id: Int, name: String, isSecret: Int? = null) {
        val client = this.client ?: return
        val updated = GroupRepository(client).update(id, name, isSecret)
        val idx = _groups.value.indexOfFirst { it.id == id }
        if (idx >= 0) {
            _groups.value = _groups.value.toMutableList().also { it[idx] = updated }
        }
    }

    suspend fun deleteGroup(id: Int) {
        val client = this.client ?: return
        GroupRepository(client).delete(id)
        _groups.value = _groups.value.filterNot { it.id == id }
        _collapsedGroupIds.value = _collapsedGroupIds.value - id
    }

    suspend fun refreshGroup(id: Int): RefreshResult {
        val client = this.client ?: throw ApiException.InvalidUrl()
        return GroupRepository(client).refresh(id)
    }

    suspend fun reorderGroups(from: Int, to: Int) {
        val client = this.client ?: return
        _groups.value = reorderAndPersist(_groups.value, from, to, { it.id }) { ids ->
            GroupRepository(client).reorder(ids)
        }
    }
}
